File: backend/panel/action_rules.py
"""实体 ID 的合法性规则，以及控件动作用到的少量口径常量。

口径只在这里写一份：校验层（`schema.py`）与运行期都从这里取，避免同一规则写两份。
本模块不依赖 pydantic，也不碰数据库，因此可以被任意层安全导入。

**不要再往这里加动作判定**：动作类型与弹窗来源的校验由 `schema.py` 的 pydantic
`Literal` 承担。同名逻辑的 **JS 版仍在用**
（`frontend/static/shared/action-rules.js` 被 `editor/home.js` 引用），这是两侧的漂移，记在 B43。
"""
from __future__ import annotations

import re

# Home Assistant 原生实体 ID：域 + 点 + 对象 ID，两段都只允许小写字母、数字与下划线，**两段各限 200**
# （与 interaction3d/config.py 的实体校验同一个数）。没有上限时 `"a" * 一千万 + ".b"` 这种「合法实体 ID」
# 会被静默收进可见范围、订阅集与日志；只限对象 ID 那一段时，域那一段仍可以无限长。
ENTITY_ID = re.compile(r"^[a-z0-9_]{1,200}\.[a-z0-9_]{1,200}$")
# 渲染器自造的虚拟实体：在原生格式前多一段固定的 virtual 前缀，
# 例如 virtual.light.abc，用于把同一个 HA 实体在不同页面作用域下区分开。
VIRTUAL_ENTITY_ID = re.compile(r"^virtual\.[a-z0-9_]{1,200}\.[a-z0-9_]{1,200}$")

# more-info 弹窗的三种来源：custom 打开组合弹窗，entity 打开指定实体的原生弹窗，
# current 打开控件自身绑定实体的弹窗。`schema.py` 校验 popupSource 时用它。
POPUP_SOURCES = frozenset({"custom", "entity", "current"})
# 这些域下的实体「开关」语义成立，才允许把 toggle 动作挂上去；
# 列表之外的域（例如 sensor、camera）只能做 more-info 或跳转。
TOGGLE_ENTITY_DOMAINS = frozenset(
    {
        "fan",
        "cover",
        "light",
        "button",
        "remote",
        "script",
        "switch",
        "climate",
        "automation",
        "media_player",
        "water_heater",
        "input_boolean",
    }
)


def valid_ha_entity_id(value: str) -> bool:
    """判断是否为 Home Assistant 原生实体 ID（不接受虚拟实体）。"""
    return bool(ENTITY_ID.fullmatch(value))


def valid_entity_id(value: str) -> bool:
    """判断是否为合法实体 ID：Home Assistant 原生 ID 或渲染器的作用域虚拟 ID。"""
    return valid_ha_entity_id(value) or bool(VIRTUAL_ENTITY_ID.fullmatch(value))

File: frontend/static/shared/action-rules.js
/**
 * 组件动作规则（Action Rule）的取值与合法性判定。
 *
 * 编辑器「动作」检查器与展示页运行时的纯逻辑模块：定义动作类型 / 弹窗来源 / 可开关实体域的
 * 允许取值，并校验动作配置与组件绑定是否匹配。不碰 DOM 也不请求后端；校验所需的页面路径集合、
 * 弹窗 ID 集合由调用方传入，传入 null 表示「暂不校验该维度」。
 */
import { isVirtualEntityId } from "./virtual-entities.js?v=2609221053";
// 「按 ID 切域」只有一份实现，走 utils/entities.js 的 entityDomainFromId。
import { entityDomainFromId } from "../utils/entities.js?v=2609221053";

// 动作类型固定三种：开关、打开更多信息、跳转页面。
export const ACTION_TYPES = Object.freeze(["toggle", "more-info", "navigate"]);

// 「更多信息」弹窗的数据来源：当前实体 / 指定实体 / 指定弹窗。
const POPUP_SOURCES = Object.freeze(["current", "entity", "custom"]);

// 这些 HA 域里的实体调用 toggle 才有意义；虚拟实体单独放行（见下方判定）。
export const TOGGLE_ENTITY_DOMAINS = new Set([
  "automation",
  "button",
  "climate",
  "cover",
  "fan",
  "input_boolean",
  "light",
  "media_player",
  "remote",
  "script",
  "switch",
  "water_heater"
]);

/**
 * 读取动作配置里的弹窗来源，做白名单收敛。
 */
function actionPopupSource(action) {
  // 旧文档里没有 popupSource 字段，默认按「当前实体」处理，保证向后兼容。
  const popupSource = String(action?.data?.popupSource || "current");
  if (POPUP_SOURCES.includes(popupSource)) {
    return popupSource;
  } else {
    // 出现未知取值时统一回退，避免把脏数据带进运行时。
    return "current";
  }
}

/**
 * 把动作配置里的弹窗相关字段归一成运行时可直接使用的结构。
 */
export function actionPopupData(actionSpec) {
  return {
    source: actionPopupSource(actionSpec),
    entityId: String(actionSpec?.data?.entityId || ""),
    popupId: String(actionSpec?.data?.popupId || "")
  };
}

/**
 * 判断动作是否依赖「当前组件所绑定的实体」。
 */
export function actionNeedsCurrentEntity(actionRule) {
  return (
    actionRule?.type === "toggle" ||
    (actionRule?.type === "more-info" && actionPopupSource(actionRule) === "current")
  );
}

/**
 * 判断给定实体 ID 是否支持 toggle 动作。
 */
export function entityIdSupportsToggle(entityId) {
  const normalizedEntityId = String(entityId || "");
  // 域前缀统一走 utils/entities.js 的 entityDomainFromId；实体名里再出现点号也不会干扰判断。
  return (
    isVirtualEntityId(normalizedEntityId) ||
    TOGGLE_ENTITY_DOMAINS.has(entityDomainFromId(normalizedEntityId))
  );
}

/**
 * 校验组件的动作配置是否成立。
 */
export function componentActionIsSupported(
  component,
  actionConfig,
  { pagePaths: pagePaths = null, popupIds: popupIds = null } = {}
) {
  // presence-sensor 是只读的在场探测组件，任何动作都不挂。
  if (!ACTION_TYPES.includes(actionConfig?.type) || component?.type === "presence-sensor") {
    return false;
  }
  const boundEntityId = String(component?.bindings?.entity?.entityId || "");
  if (actionConfig.type === "toggle") {
    return entityIdSupportsToggle(boundEntityId);
  }
  if (actionConfig.type === "navigate") {
    // 目标页面必须存在；pagePaths 为 null 时只要求非空，多为编辑器加载早期。
    const target = String(actionConfig.target || "");
    return !!target && (!pagePaths || pagePaths.has(target));
  }
  const resolvedPopupSource = actionPopupSource(actionConfig);
  if (resolvedPopupSource === "current") {
    return !!boundEntityId;
  }
  if (resolvedPopupSource === "entity") {
    return !!actionConfig.data?.entityId;
  }
  // 走到这里只剩 custom：弹窗必须存在，popupIds 为 null 时只要求填了 ID。
  const popupId = String(actionConfig.data?.popupId || "");
  return !!popupId && (!popupIds || popupIds.has(popupId));
}
